### DISCUSSION TITLE
Multiple USB CDC - async write issues

### DISCUSSION THREAD

**Discussion by mishal** [micropython/micropython#13040] [RP2040 / Pico] [answered]
Hello community,

I'm working on a project which requires multiple CDC interfaces. I have succesfully patched micropython rp2 port to setup 4 additional interfaces on Raspberry Pico and implemented C module (ttyacm, based on original code found on RPI forum - author hippy) which calls underlaying tiny usb library. Still wip, but so far so good. 

I have an example test which writes some bytes asynchronously using `asycio.StreamWriter` to the interface and runs forever. It works fine with micropython 1.21.0. I get the output on the other interface. I tried this with latest upy code, but it does not work as expected. The async script hangs on the first "flushed" (see below). When I connect to the CDC it writes another line, but hangs after sending it. I have to disconnect from the second serial device and then connect again... simple non-async version works fine (see below)

I used git bisect to find the first commit which breaks the funcionality and found commit: 8c432ea2d437255eefb4aff18764936ad11b70c7 

This commit is: `rp2: Remove 1ms timeout to make idle waiting tickless.
I do not understand what is going on under the hood... Can someone shed some light on this?  Thank you.

main.py:

```
import asyncio
import ttyacm

async def write_msg(aw, msg):
    aw.write(msg)

    print("waiting to flush")
    await aw.drain()
    print("flushed")


async def writer_example():
    itf = ttyacm.TTYACM(4)
    aw = asyncio.StreamWriter(itf)
    i = 0
    while True:
        i += 1
        await write_msg(aw, b"test" + str(i).encode())        
        #print("written", itf.write(b"test"))        
        await asyncio.sleep(1)

async def main():
    asyncio.create_task(writer_example())
    while True:
        await asyncio.sleep(10000)

loop = asyncio.get_event_loop()

try:
    print("RUNNING MAIN")
    asyncio.run(main())
except KeyboardInterrupt:
    print("Interrupted by CTRL+C")
finally:
    asyncio.new_event_loop()

```
When I try to use simple non-async version, it works as expected...

```
import time
import ttyacm

itf = ttyacm.TTYACM(4)

while True:
    print("test")
    itf.write(b"test")
    time.sleep(1)

```

**Comment by projectgus** [comment]
> I used git bisect to find the first commit which breaks the funcionality and found commit: https://github.com/micropython/micropython/commit/8c432ea2d437255eefb4aff18764936ad11b70c7
>  I do not understand what is going on under the hood... Can someone shed some light on this? Thank you.

Hi @mishal,

The rp2 port regularly pauses by issuing the WFE (wait for event) instruction, which suspends the CPU until an interrupt or other event triggers.

Prior to the linked commit, the rp2 port would receive an interrupt at least every 1ms to wake up the CPU, at which time the MicroPython runtime might do some other processing (like calling scheduled functions, for example). This is very similar to a "tick" interrupt [^].

After the linked commit, this 1ms wakeup no longer happens. If there isn't an interrupt to trigger progress, the CPU stays asleep ("tickless"). To work, any driver will need to ensure that there's an interrupt (or a WFE-compatible Event) whenever it has to progress its state, it can no longer rely on the MicroPython runtime waking up every millisecond and potentially polling it.

Does that answer your question?

[^] In the case of the rp2 port it's not exactly a tick interrupt because it doesn't happen every 1ms, instead the CPU stops executing for at most 1ms. But it's similar to one, for the purposes of this question.

**Reply by mishal** [reply]
@projectgus yes, thank you, that answers my question. 

Should I trigger some event/s in my code to wake up the CPU? How?

While trying to debug my issue I just tested this simple asyncio program which **does not work** on [v1.22.0-preview.164.gfce8d9fd5 (2023-11-20) .uf2](https://micropython.org/resources/firmware/RPI_PICO-20231120-v1.22.0-preview.164.gfce8d9fd5.uf2) **but works on** [v1.21.0 (2023-10-05) .uf2](https://micropython.org/resources/firmware/RPI_PICO-20231005-v1.21.0.uf2). Is this a bug or am I doing something wrong?

```
import asyncio

async def write_msg():    
    print("waiting to flush")
    await asyncio.sleep(2)
    print("flushed")

async def writer_example():
    while True:
        await write_msg()
        await asyncio.sleep(1)

async def main():
    asyncio.create_task(writer_example())
    while True:
        await asyncio.sleep(10000)

loop = asyncio.get_event_loop()

try:
    print("RUNNING MAIN")
    asyncio.run(main())
except KeyboardInterrupt:
    print("Interrupted by CTRL+C")
finally:
    asyncio.new_event_loop()

```

**Reply by projectgus** [reply]
@mishal Can confirm this freezes for me as well. This is definitely a bug!

It feels like a variant of https://github.com/micropython/micropython/issues/12925

Apologies, I should have twigged when I replied to you the first time.

**Reply by mishal** [reply]
@projectgus ok, thank you. I'll be watching #12925

**Reply by mishal** [reply, marked as answer]
@projectgus thanks, was fixed by merging #13096
